integration-tests/msa/createMsa.test.ts

```diff
@@ -1,69 +1,244 @@
-        assert.notEqual(chainEvents["transactionPayment.TransactionFeePaid"], undefined);
+        keys = await createAndFundKeypair();
     });
 
-    it("should successfully mimic a user's path using tokens", async function () {
-        const keys = (await createAndFundAccount(fundingInputs)).newAccount;
-        const createMsaEvents = await createMsa(api, keys);
-        const msaId = createMsaEvents["msa.MsaCreated"][0]
-        assert.notEqual(msaId, undefined);
-
-        const newKeys: KeyringPair = createAccount();
-        const payload = { msaId: msaId, newPublicKey: newKeys.publicKey, expiration: 50 }
-
-        const addKeyData = api.registry.createType("PalletMsaAddKeyData", payload);
+    describe("createMsa", function () {
 
-        const ownerSig = signPayloadSr25519(keys, addKeyData);
-        const newSig = signPayloadSr25519(newKeys, addKeyData);
+        it("should successfully create an MSA account", async function () {
+            const f = ExtrinsicHelper.createMsa(keys);
+            const chainEvents = await f.fundAndSend();
+            const msaCreatedEvent = chainEvents["msa.MsaCreated"];
 
-        const publicKeyEvents = await addPublicKeyToMsa(api, keys, ownerSig, newSig, payload);
+            assert.notEqual(chainEvents["system.ExtrinsicSuccess"], undefined, "should have returned an ExtrinsicSuccess event");
+            assert.notEqual(msaCreatedEvent, undefined, "should have returned  an MsaCreated event");
+            assert.notEqual(chainEvents["transactionPayment.TransactionFeePaid"], undefined, "should have returned a TransactionFeePaid event");
 
-        assert.notEqual(publicKeyEvents["msa.PublicKeyAdded"], undefined, 'should have added public key');
+            if (ExtrinsicHelper.api.events.msa.MsaCreated.is(msaCreatedEvent)) {
+                msaId = msaCreatedEvent.data.msaId;
+            }
+        });
 
-        const deleteEvents = await deletePublicKey(api, keys, newKeys.publicKey);
-
-        assert.notEqual(deleteEvents["msa.PublicKeyDeleted"], undefined, 'should have deleted public key');
+        it("should fail to create an MSA for a keypair already associated with an MSA", async function () {
+            const op = ExtrinsicHelper.createMsa(keys);
+            await assert.rejects(op.fundAndSend(), {
+                name: 'KeyAlreadyRegistered',
+            });
+        });
+    });
 
-        const createSchemaEvents = await createSchema(api, keys, AVRO_GRAPH_CHANGE, "AvroBinary", "OnChain");
-        assert.notEqual(createSchemaEvents["schemas.SchemaCreated"], undefined, 'should have created schema');
+    describe("addPublicKeyToMsa", function () {
+        let badKeys: KeyringPair;
+        let payload: AddKeyData = {};
+        let ownerSig: Sr25519Signature;
+        let newSig: Sr25519Signature;
+        let badSig: Sr25519Signature;
+        let addKeyData: Codec;
+
+        before(async function () {
+            authorizedKeys.push(await createAndFundKeypair());
```

Comment:
It doesn't seem like we more than one of these keypairs. Why do we need this array?

Reply:
Any test that (successfully) adds an authorized keypair pushes it to this array. This way, later on when we test that we cannot retire an MSA until all authorized keys have been deleted, we can simply go through the array and delete any keys that were authorized, thus allowing us to successfully retire the MSA.

Reply:
it doesn't look like we access any keys apart from the first. am i mistaken?

Reply:
Look on line 176... we push a second keypair

Reply:
Also any future test cases that add keypairs to that MSA can push to that array and the later tests will not need to be modified.

Reply:
> Look on line 176... we push a second keypair

I see that, but I don't see us accessing it anywhere. If we're not accessing it.... do we need it?

Reply:
See here: https://github.com/LibertyDSNP/frequency/pull/850#discussion_r1061842811